mkosi/machine.py

```diff
@@ -101,19 +89,34 @@ def __enter__(self) -> Machine:
 
             cmd = " ".join(str(x) for x in cmdline)
 
-            # Here we have something equivalent to the command lines used on spawn() and run() from backend.py
-            # We use pexpect to boot an image that we will be able to interact with in the future
-            # Then we tell the process to look for the # sign, which indicates the CLI for that image is active
-            # Once we've build/boot an image the CLI will prompt "root@image ~]# "
-            # Then, when pexpects finds the '#' it means we're ready to interact with the process
+            # Here we have something equivalent to the command lines used on spawn() and run() from backend.py.
+            # We use pexpect to boot an image that we will be able to interact with in the future.
+            # Then we tell the process to look for the # sign, which indicates the CLI for that image is active.
+            # Once we've build/boot an image the CLI will prompt "root@image ~]# ".
+            # Then, when pexpects finds the '#' it means we're ready to interact with the process.
             self._serial = pexpect.spawnu(cmd, logfile=None, timeout=240)
             self._serial.expect("#")
             self.stack = stack.pop_all()
 
+    def __enter__(self) -> Machine:
+        if self.args.verb in MKOSI_COMMANDS_SUDO:
+            check_root()
+            unlink_output(self.args)
+
+        if self.args.verb == Verb.build:
+            check_output(self.args)
+
+        if needs_build(self.args):
+            check_root()
+            check_native(self.args)
+            init_namespace(self.args)
+            build_stuff(self.args)
+        self._boot()
+
         return self
```

Comment:
This has unnecessarily been moved downwards. You should always try to avoid unnecessary changes in PRs to keep them small.

Reply:
Hmm, that's fair. Didn't really think of that until now, will keep it in mind in the future!